src/analyzing/CFG_build.ml

```diff
@@ -240,41 +259,122 @@ let rec cfg_stmt : state -> F.nodei option -> stmt -> cfg_stmt_result =
-  | Try (try_st, catches, finally_st) ->
+  | Try (try_st, catches, else_st, finally_st) ->
       (* previ ->
        * newi ->
        * try -> catchesi --> catch1 -|
        *                 |->  ...   -|
        *                 |-> catchN -|
-       *                 |-----------|-> newfakefinally -> finally
-       *
+       *                 |-----------|
+       *     -> elsei    --> else    |-> newfakefinally -> finally
        *)
       let newi = state.g#add_node { F.n = NOther (Noop "try") } in
-      state.g |> add_arc_from_opt (previ, newi);
       let catchesi = state.g#add_node { F.n = NOther (Noop "catch") } in
-      let state' = { state with try_catches_opt = Some catchesi } in
-      let finaltry = cfg_stmt_list state' (Some newi) try_st in
-      state.g |> add_arc_from_opt (finaltry, catchesi);
+      let elsei = state.g#add_node { F.n = NOther (Noop "else") } in
       let newfakefinally = state.g#add_node { F.n = NOther (Noop "finally") } in
+
+      (* In the try/except/else clauses, if there is a return, then
+       * execute the finally clause, or if there's no finally clause, propagate
+       * the return.
+       *)
+      let return_destination =
+        match finally_st with
+        | [] -> state.return_destination
+        | _ -> newfakefinally
+      in
+
+      (* From prev to try. *)
+      state.g |> add_arc_from_opt (previ, newi);
+
+      (* Inside try may go to catches. *)
+      let try_state =
+        { state with throw_destination = Some catchesi; return_destination }
+      in
+      let finaltry = cfg_stmt_list try_state (Some newi) try_st in
+
+      (* The following will help determine whether we should exit
+       * or continue the flow after the whole statement
+       *)
+      (* When this is true, the flow stops after the finally clause. *)
+      let try_must_return =
+        match finaltry with
+        | None -> !(state.may_return) && not !(state.may_throw)
+        | _ -> false
+      in
+      (* When this is true, control must go to the else clause. *)
+      let try_must_not_exit =
+        match finaltry with
+        | None -> false
+        | _ -> (not !(state.may_throw)) && not !(state.may_return)
+      in
+
+      (* From end of try to else. *)
+      state.g |> add_arc_from_opt (finaltry, elsei);
+
+      (* In the else clause, if an exception is thrown, go to the finally
+       * clause, or if there is no finally clause, propagate the exception.
+       *)
+      let throw_destination =
+        match finally_st with
+        | [] -> state.throw_destination
+        | _ -> Some newfakefinally
+      in
+      let else_state = { state with throw_destination; return_destination } in
+      let finalelse = cfg_stmt_list else_state (Some elsei) else_st in
+
+      (* If the else clause is reached, and this is true, then the flow stops
+       * after the finally clause.
+       *)
+      let else_must_exit =
+        match finalelse with
+        | None -> true
```

Comment:
Don't we need to check `may_return` here?

Reply:
This logic will be removed to simplify this PR.